Add Modifications and RootFeature

s9latimm opened the pull request:
Add modifications to test broken FMs, refactor root nodes and clean up public interface.

- [x] Add removal of edges, children and features as modifications
  - [x] Enable creation of broken FM through inheritance
- [x] Remove `Feature::isRoot` to prevent misconception
  - [x] Add `RootFeature` class
  - [x] Refactor builder
  - [x] Refactor comparisons to root with `llvm::isa`
- [x] Clean up unused / deprecated constructors
- [x] Update python bindings
  - [x] Add `RootFeature`
  - [x] Add `feature_model.get_feature`
  - [x] Clean up modifications / deprecated methods
- [x] add `venv/` to gitignore

s9latimm committed:
Add modifications and clean up public interface

M	.gitignore
M	bindings/python/tests/feature/test_feature.py
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/pybind_Feature.cpp
M	bindings/python/vara-feature/pybind_FeatureModel.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Revert binding

M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/pybind_Feature.cpp

s9latimm committed:
Move parent test

M	bindings/python/tests/feature_model/test_feature_model.py

s9latimm committed:
Check root

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Fix ASAN prevents SEGV

M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Clean up

M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelTransaction.cpp

s9latimm commented on bindings/python/vara-feature/pybind_Feature.cpp:
```diff
@@ -80,13 +81,6 @@ void init_feature_module_binary_feature(py::module &M) {
       .def(py::init<std::string, bool>())
       .def(py::init<std::string, bool,
                     std::vector<vara::feature::FeatureSourceRange>>())
-      .def(py::init<std::string, bool,
-                    std::vector<vara::feature::FeatureSourceRange>,
-                    vara::feature::Feature *>())
-      .def(py::init<std::string, bool,
-                    std::vector<vara::feature::FeatureSourceRange>,
-                    vara::feature::Feature *,
-                    std::vector<vara::feature::FeatureTreeNode *>>())
```
@padupr do you need these constructors?

vulder replied:
nope, je does not need them

padupr replied:
I don't think so.

s9latimm committed:
WIP: Add RootFeature

M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp
A	unittests/Feature/RootFeature.cpp

s9latimm committed:
WIP: Enable tests

M	include/vara/Feature/FeatureModel.h
M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Use isa

M	bindings/python/tests/feature/test_feature.py
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/pybind_Feature.cpp
M	unittests/Feature/FeatureModelTransaction.cpp

vulder commented on bindings/python/vara-feature/pybind_Feature.cpp:
```diff
@@ -45,6 +42,10 @@ void init_feature_module_feature(py::module &M) {
                              R"pbdoc(The name of the feature.)pbdoc")
+      .def(
+          "is_root",
+          [](vf::Feature &F) { return llvm::isa<vf::RootFeature>(F); },
```
```suggestion
          [](const vf::Feature &F) { return llvm::isa<vf::RootFeature>(F); },
```
I, think this should be const, otherwise, we could get a problem when we call python functions on an underlying const object.

vulder commented on bindings/python/vara-feature/pybind_FeatureModel.cpp:
```diff
@@ -22,6 +22,12 @@ void init_feature_model_module(py::module &M) {
            R"pbdoc(Returns the root Feature.)pbdoc")
+      .def(
+          "get_feature",
+          [](vf::FeatureModel &FM, const std::string &Name) {
```
Do we get a ref here from the python side? Not sure, normally, if we would call a member function on an C++ object this would be a pointer.

s9latimm replied:
So far we always used refs as arguments and looking at pybind11s documentation, they are doing so as well -- idk.
It seems like ptrs work too, at least they do not produce errors yet.

vulder replied:
ok then let's stay with refs

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -278,6 +257,21 @@ class NumericFeature : public Feature {
+
+class RootFeature : public Feature {
+public:
+  RootFeature(string Name = "root")
```
```suggestion
  explicit RootFeature(string Name = "root")
```
Ctors with only one parameter should be explicit to prevent unwanted implicit conversions.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -438,16 +446,17 @@ template <typename... Rules> class FeatureModelConsistencyChecker {
-      return F->isRoot() ||
-             // Every parent of a Feature needs to have the Feature as a child.
+      return llvm::isa<RootFeature>(F) ||
+             // Every parent of a Feature needs to have that as a child.
```
```suggestion
             // Every parent of a Feature needs to have it as a child.
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -438,16 +446,17 @@ template <typename... Rules> class FeatureModelConsistencyChecker {
     return std::all_of(FM.begin(), FM.end(), [](Feature *F) {
-      return F->isRoot() || (F->getParent() != nullptr);
+      return llvm::isa<RootFeature>(F) ||
+             (F->getParent() && F->getParentFeature());
```
Why do we need the extra `getParentFeature` check?

s9latimm replied:
In this loop we only iterate over feature nodes, so relationships parents are not covered. If we assert, that each feature has an ancestor, this implies that also every relationship has a parent of some kind (assuming we prohibit directly nested relationships).

vulder replied:
But how does checking if the parent is a Feature help if we don't iterate over the other nodes?

s9latimm replied:
If we have a feature in a relationship, the parent node will only be of that kind. So if we find an parent feature (which can't be the actual parent), this means, that the relationship node itself has to have a parent, which again should be a feature. So we implicitly check if every relationship has a parent as well.

tl;dr: If we check that features are connected to a parent feature, this may be transitive over a relationship -- so the relationship itself has to have a parent in the first place.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -164,15 +166,15 @@ class FeatureModelBuilder : private FeatureModel {
             typename = typename std::enable_if_t<
                 std::is_base_of_v<Feature, FeatureTy>, int>>
-  Feature *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
+  FeatureTy *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
     if (!Features
              .try_emplace(FeatureName,
                           std::make_unique<FeatureTy>(
                               FeatureName, std::move(FurtherArgs)...))
```
```suggestion
  FeatureTy *makeFeature(const std::string &FeatureName, Args &&...FurtherArgs) {
    if (!Features
             .try_emplace(FeatureName,
                          std::make_unique<FeatureTy>(
                              FeatureName, std::forward<Args>(FurtherArgs)...))
```

s9latimm replied:
Moving `FeatureName` during the creation breaks, as it is needed for the emplacement afterwards.

vulder replied:
ok I remove the move

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -164,15 +166,15 @@ class FeatureModelBuilder : private FeatureModel {
   template <typename FeatureTy, typename... Args,
             typename = typename std::enable_if_t<
                 std::is_base_of_v<Feature, FeatureTy>, int>>
```
```suggestion
  template <typename FeatureTy, typename... Args,
            std::enable_if_t<std::is_base_of_v<Feature, FeatureTy>, int> = 0>
```

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -22,7 +22,8 @@ class OrderedFeatureVector {
   OrderedFeatureVector(std::initializer_list<Feature *> Init) { insert(Init); }
   template <class FeatureIterTy>
   OrderedFeatureVector(FeatureIterTy Begin, FeatureIterTy End) {
```
```suggestion
  OrderedFeatureVector(FeatureIterTy &&Begin, FeatureIterTy &&End) {
```

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -42,7 +49,8 @@ class OrderedFeatureVector {
 
   template <class FeatureIterTy>
   void insert(FeatureIterTy Begin, FeatureIterTy End) {
```
```suggestion
  void insert(FeatureIterTy &&Begin, FeatureIterTy &&End) {
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -164,15 +166,15 @@ class FeatureModelBuilder : private FeatureModel {
             typename = typename std::enable_if_t<
                 std::is_base_of_v<Feature, FeatureTy>, int>>
-  Feature *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
+  FeatureTy *makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
```
```suggestion
  FeatureTy *makeFeature(std::string FeatureName, Args... FurtherArgs) {
```

vulder commented on unittests/Feature/BinaryFeature.cpp:
```diff
@@ -44,9 +42,7 @@ TEST(BinaryFeature, BinaryFeatureChildren) {
-  if (auto *F =
-          llvm::dyn_cast<vara::feature::Feature>(*FM->getFeature("a")->begin());
-      F) {
+  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin()); F) {
```
```suggestion
  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin())) {
```
If you don't have a more complex check of need the variable defined in the else branch you can just use an normal if here without initializer.

vulder commented on unittests/Feature/FeatureModel.cpp:
```diff
@@ -45,20 +34,19 @@ TEST(FeatureModel, cloneUnique) {
-              testing::KilledBySignal(SIGSEGV), ".*");
-  EXPECT_TRUE(Clone->getFeature("a")->isRoot());
+  EXPECT_DEATH(EXPECT_TRUE(FM->getFeature("a")->getParent()), ".*");
+  EXPECT_FALSE(Clone->getFeature("a")->getParent());
```
Shouldn't this become a `isa` check for RootNode instead of an empty parent check?

s9latimm replied:
Indeed.

vulder commented on unittests/Feature/NumericFeature.cpp:
```diff
@@ -60,9 +58,7 @@ TEST(NumericFeature, NumericFeatureChildren) {
-  if (auto *F =
-          llvm::dyn_cast<vara::feature::Feature>(*FM->getFeature("a")->begin());
-      F) {
+  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin()); F) {
```
```suggestion
  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin())) {
```

vulder commented on unittests/Feature/RootFeature.cpp:
```diff
@@ -0,0 +1,54 @@
+  EXPECT_EQ(
+      std::distance(FM->getFeature("a")->begin(), FM->getFeature("a")->end()),
+      1);
+  if (auto *F = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin()); F) {
+    EXPECT_EQ("aa", F->getName());
+  } else {
+    FAIL();
+  }
```
```suggestion
  Feature *FirstChild = llvm::dyn_cast<Feature>(*FM->getFeature("a")->begin());
  EXPECT_NE(FirstChild, nullptr);
  EXPECT_EQ("aa", F->getName());
```
This produces more precise error messages.

s9latimm replied:
I agree, but Clang-Tidy wants me to assure the pointer cannot be null in this block -- not even `ASSERT_` satisfies here.

vulder replied:
Does a assert(FirstChild) help?

s9latimm replied:
Yes, but `Release` will still fail.

vulder replied:
ok, your right, then let's keep this.

s9latimm committed:
Apply suggestions from code review

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

s9latimm committed:
Apply suggestions from code review

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

s9latimm committed:
Apply suggestions from code review

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/RootFeature.cpp

s9latimm committed:
Apply suggestions from code review

M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Tidy

M	unittests/Feature/RootFeature.cpp

s9latimm committed:
Every feature should have a feature parent

M	include/vara/Feature/FeatureModel.h
